feat: add WELLDONE extension wallet

Yoon-Suji opened the pull request:
# Description

WELLDONE extension wallet is multi-chain wallet that support NEAR protocol. You can read more information in [here](https://docs.welldonestudio.io/wallet).

# Checklist:
<!-- CHECKLIST_TYPE: ALL -->
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
<!-- /CHECKLIST_TYPE -->

# Type of change.
<!-- CHECKLIST_TYPE: ONE -->
- [ ] FIX - a PR of this type patches a bug.
- [x] FEATURE - a PR of this type introduces a new feature.
- [ ] BUILD - a PR of this type introduces build changes.
- [ ] CI - a PR of this type introduces CI changes.
- [ ] DOCS - a PR of this type introduces DOCS improvement.
- [ ] STYLE - a PR of this type introduces style changes.
- [ ] REFACTOR - a PR of this type introduces refactoring.
- [ ] PERFORMANCE - a PR of this type introduces performance changes.
- [ ] TEST - a PR of this type adds more tests.
- [ ] CHORE - a PR introduces other changes than the specified above.
<!-- /CHECKLIST_TYPE -->

kujtimprenkuSQA commented:
Hey, @Yoon-Suji thank you for submitting this PR.

Could you please add a (docs) `Readme.md` file to the package like it's done with other packages (wallets) in this repo?

Before reviewing the code I tried to install and test out the functionality of this wallet/extension but for some reason, I am unable to create a new wallet with this extension it always gets stuck in this view:

![create-new-wallet](https://user-images.githubusercontent.com/95851345/193262301-73031578-3447-43b5-b81c-143bb3267c8e.PNG)

Yoon-Suji commented:
@kujtimprenkuSQA Hi, I'm sorry for late response😂
I tried it again and it works well. Can you please check the chrome version? And also I added docs in packages.

kujtimprenkuSQA commented:
This is the chrome version I am using on this machine, I think it's the latest version:

**Version 106.0.5249.91 (Official Build) (64-bit)**

The Welldone wallet is still not working for me, will ask other team members to have a look at it too.

amirsaran3 commented:
Hey @Yoon-Suji. I was able to successfully create an account using the WELLDONE extension. I built your branch to test the functionality and I am seeing a couple of bugs:

1. When refreshing page I get logged out (session is cleared).
2. `signAndSendTransaction` and `signAndSendTransactions` functions are not working for me. I am getting this error which is not letting me to sign transactions: 
![image](https://user-images.githubusercontent.com/95852708/193796454-4524f465-785a-4a8f-b306-430cd29d28aa.png)

amirsaran3 commented on packages/welldone-wallet/src/lib/welldone.ts:
```diff
@@ -0,0 +1,328 @@
+      };
+    },
+
+    async signAndSendTransaction({ signerId, receiverId, actions }) {
```
You should consider using the `signTransactions` function in `packages\wallet-utils\src\lib\sign-transactions.ts` when signing to avoid using these "helper function" like `_validateAccessKey`, `convertTransaction` and `convertActions`. All of the above functions are used in `signTransactions`. This way we avoid duplicating code. 

You will have to provide a `signer: Signer` object to the `signTransactions` function. You can check out the `nightly` package and see how they have done it.

Yoon-Suji replied:
We implemented a [dapp:sendTransaction](https://docs.welldonestudio.io/wallet/developer-guide/provider-api/method#sendTransaction) method of signing before sending a transaction on its own inside the wallet. So we use it instead of using the `signTransactions` function.

Yoon-Suji replied:
Hey, I changed the `signAndSendTransaction` method to using `signTransactions` function. But the `signAndSendTransactions` method still uses the `dapp:sendTransaction` method (the wallet signs and sends transactions) because we do not currently support signing multiple transactions. This code will change after the wallet has been updated.

DamirSQA replied:
Hi @Yoon-Suji , I'm unable to create or import wallet as well. Getting stuck on "Create a wallet" screen
![image](https://user-images.githubusercontent.com/111346686/196677135-709b514a-088e-4c30-9648-d292e1afef56.png)

Tried it on two machines with Chrome Versions 106.0.5249.119 (Official Build) (64-bit) and 106.0.5249.103 (Official Build) (64-bit). The same goes for Brave browser so you might want to investigate. 

Small thing I've noticed with password rule text... It says "write more than 8 letters" instead of "write 8 or more characters" if you want to change that.

Yoon-Suji replied:
Hi @DamirSQA , We are trying to reproduce the exceptional situation where the error occurs. 
So I would really appreciate it if you could send me the console log by right-clicking on the `Create a Wallet` screen and clicking the inspect button.
Thanks for reporting the issue.

DamirSQA replied:
You're welcome, here is the log

![image](https://user-images.githubusercontent.com/111346686/196904132-f85caaf3-a347-4fa9-a62e-758e0b6ffd5f.png)

Yoon-Suji replied:
Thanks, @DamirSQA Could you turn off all the other wallet extensions and try again?

DamirSQA replied:
Removing extensions didn't make a difference.  On some devices it does work, on the others does not.

amirsaran3 commented on packages/welldone-wallet/src/lib/welldone.ts:
```diff
@@ -0,0 +1,328 @@
+
+async function initWalletState(): Promise<WelldoneWalletState> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const wallet = (window as any).dapp;
```
Instead of using `(window as any).dapp` can you declare something like this at the top of the file: 

```ts
declare global {
  interface Window {
    dapp: WalletProvider;
  }
}
```

Yoon-Suji replied:
ok I'll fix it!

AmmarHumackicSQA commented:
Hi @Yoon-Suji , thanks for submitting the PR. As a reminder, the security statement needs to be shown either on the GitHub profile/repo or the website. You can check it [here](https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#:~:text=GitHub%20account%20or%20Website) again. Thanks!

Yoon-Suji commented:
Hi @AmmarHumackicSQA 
We added the security statement to our GitHub organization profile! you can check it [here](https://github.com/welldonestudio)

Yoon-Suji commented:
Hi @amirsaran3 
First of all, thanks for letting us know about the bug.
There was an issue with the RPC node we use, which is why the `signAndSendTransaction` and `signAndSendTransactions methods` failed. We fixed it and uploaded the updated wallet to the chrome webstore. Could you please delete the current version of WELLDONE extension wallet and [re-download](https://chrome.google.com/webstore/detail/welldone-wallet-for-multi/bmkakpenjmcpfhhjadflneinmhboecjf) it? 
And I fixed some of the code by reflecting on your comments. Please check it!

amirsaran3 commented:
> Hi @amirsaran3 First of all, thanks for letting us know about the bug. There was an issue with the RPC node we use, which is why the `signAndSendTransaction` and `signAndSendTransactions methods` failed. We fixed it and uploaded the updated wallet to the chrome webstore. Could you please delete the current version of WELLDONE extension wallet and [re-download](https://chrome.google.com/webstore/detail/welldone-wallet-for-multi/bmkakpenjmcpfhhjadflneinmhboecjf) it? And I fixed some of the code by reflecting on your comments. Please check it!

Hey @Yoon-Suji. I tested your new version and signing works. However wallets should have a function that allows us to just sign the transaction and return the signature. Is it possible to add `near:signTransaction` and `near:signTransactions` request methods?

Yoon-Suji commented:
> > Hi @amirsaran3 First of all, thanks for letting us know about the bug. There was an issue with the RPC node we use, which is why the `signAndSendTransaction` and `signAndSendTransactions methods` failed. We fixed it and uploaded the updated wallet to the chrome webstore. Could you please delete the current version of WELLDONE extension wallet and [re-download](https://chrome.google.com/webstore/detail/welldone-wallet-for-multi/bmkakpenjmcpfhhjadflneinmhboecjf) it? And I fixed some of the code by reflecting on your comments. Please check it!
> 
> Hey @Yoon-Suji. I tested your new version and signing works. However wallets should have a function that allows us to just sign the transaction and return the signature. Is it possible to add `near:signTransaction` and `near:signTransactions` request methods?

Hey @amirsaran3. Welldone wallet has the `dapp:sign` method, which is the method of signing a message. I also added a `signer:Signer` object as you suggested and used it to `verifyOwner` function, please check!

amirsaran3 commented:
@Yoon-Suji After latest changes I get this error when trying to sign message:
![image](https://user-images.githubusercontent.com/95852708/195043729-5deda71b-aa99-400a-b552-5a305eb1a9ae.png)
Also there is no info about the `dapp:sign` method in your docs (https://docs.welldonestudio.io/wallet/developer-guide/provider-api/method).

Yoon-Suji commented:
@amirsaran3
This is the [issue](https://github.com/near/nearcore/issues/7769) that the old query was deprecated in `view_access_key` RPC api during the testnet upgrade process. NEAR replied that it would revert for users who use old query, but we modified the wallet in a way that does not use old query according to NEAR's initial intention. But the testnet is currently down, so if it operates again, the modifications will be reflected after the test.

And also I'll add a description of the `dapp:sign` method to our docs. Could you leave the same comment on our [Github issue](https://github.com/welldonestudio/welldonestudio.github.io/issues)? We want the docs github issue activated :)

Yoon-Suji commented:
Hi @amirsaran3 , we checked again and it seems that all works well with WELLDONE version 0.0.25 (latest version updated in the store) Could you try again? And we updated our [docs](https://docs.welldonestudio.io/wallet/developer-guide/provider-api/method/#sign) too!

Yoon-Suji commented:
hey, we implemented the `signTransaction` function.

If the `signMessage` method in Signer takes arguments as a transaction, it signs Transaction, if the method takes arguments as a message, it signs message. So I also changed `signAndSendTransaction` method to using `signTransactions` in `packages\wallet-utils\src\lib\sign-transactions.ts`

But we are not supported signing multiple transactions because currently the Welldone wallet was ready to send multiple transactions, but it is not ready to sign multiple transactions. So the `signAndSendTransactions` method still using `dapp:sendTransaction` method which is signing and sending transactions at once in the wallet. If required, we will include multiple transaction signing in the next wallet release.

DamirSQA commented on packages/welldone-wallet/src/lib/injected-welldone.ts:
```diff
@@ -0,0 +1,37 @@
+  deprecated?: boolean;
+}
+
+export interface RequestParams {
```
`RequestParams`  and `WalletProvider` are extending interfaces, so it would be nice to remove unnecessary export and put them on top of each other before `WelldoneWalletState`

```
interface RequestParams {
    ...
}

interface WalletProvider {
    ...
}

export interface WelldoneWalletState {
    ...
}

```

Yoon-Suji replied:
Thanks! I remove unnecessary export at `RequestParams`, but I left `WalletProvider` as it was since I import `WalletProvider` from `welldone.ts`
